### PR TITLE
enh (topology): Creation of a mechanism to include a replacement page in the topology

### PR DESCRIPTION
**Pull request by callapa**
## Description

The aims is to create a mechanism to replace a topology page without changing its page number.

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [x] 23.04.x (master)

#### Community contributors & Centreon team

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/www/include/common/common-Func.php [hunk 1]
```diff
@@ -1552,21 +1552,31 @@ function get_child($id_page, $lcaTStr)
...
-                ORDER BY topology_order, topology_group ";
+        $rq = <<<SQL
+            SELECT topology_parent, topology_name, topology_id, topology_url, topology_url_substitute,
+                topology_page, topology_url_opt, is_react
+            FROM topology
+            WHERE topology_parent = :id_page
+              AND topology_page IS NOT NULL
+              AND topology_show = '1'
+            ORDER BY topology_order, topology_group
+        SQL;
```
Comment (abuathier) [inline centreon/www/include/common/common-Func.php:1574]: php-cs-fixer
```suggestion
            SQL;
```

File: centreon/www/include/core/pathway/pathway.php [hunk 2]
```diff
@@ -46,12 +46,12 @@
...
      */
     $statementSelect = $pearDB->prepare(
-        'SELECT topology_url FROM topology WHERE topology_page = :topology_page'
+        'SELECT topology_url, topology_url_substitute FROM topology WHERE topology_page = :topology_page'
     );
     $statementSelect->bindValue(':topology_page', $p, \PDO::PARAM_INT);
     if ($statementSelect->execute()) {
         $result = $statementSelect->fetch(\PDO::FETCH_ASSOC);
-        if ($result !== false && $result['topology_url'] != $url) {
+        if ($result !== false && $result['topology_url'] != $url && $result['topology_url_substitute'] === null) {
```
Comment (abuathier) [inline centreon/www/include/core/pathway/pathway.php:54]: What if `topology_url_substitute = ''` (technically possible) ?

File: centreon/www/class/centreonACL.class.php [hunk 3]
```diff
@@ -1447,23 +1447,26 @@ public function queryBuilder($condition, $field, $stringlist)
...
         if ($this->admin) {
-            return 1;
+            return self::ACL_ACCESS_READ_WRITE;
         } elseif (isset($this->topology[$p])) {
-            if ($checkAction && $this->topology[$p] == 2 &&
-                isset($_REQUEST['o']) && $_REQUEST['o'] == 'a'
+            if (
+                $checkAction
+                && $this->topology[$p] == self::ACL_ACCESS_READ_ONLY
+                && isset($_REQUEST['o']) && $_REQUEST['o'] == 'a'
```
Comment (abuathier) [inline centreon/www/class/centreonACL.class.php:1463]: Strictness + idea of writing for lisibility (subjective).
```suggestion
                && ($_REQUEST['o'] ?? null) === 'a'
```

File: centreon/www/class/centreonACL.class.php [hunk 4]
```diff
@@ -1447,23 +1447,26 @@ public function queryBuilder($condition, $field, $stringlist)
...
         $this->checkUpdateACL();
         if ($this->admin) {
-            return 1;
+            return self::ACL_ACCESS_READ_WRITE;
         } elseif (isset($this->topology[$p])) {
-            if ($checkAction && $this->topology[$p] == 2 &&
-                isset($_REQUEST['o']) && $_REQUEST['o'] == 'a'
+            if (
+                $checkAction
+                && $this->topology[$p] == self::ACL_ACCESS_READ_ONLY
```
Comment (abuathier) [inline centreon/www/class/centreonACL.class.php:1462]:
```suggestion
                && $this->topology[$p] === self::ACL_ACCESS_READ_ONLY
```

File: centreon/www/class/centreonACL.class.php [hunk 5]
```diff
@@ -1447,23 +1447,26 @@ public function queryBuilder($condition, $field, $stringlist)
      *
      * @param string $p
      * @param bool $checkAction
+     *
      * @return int | 1 : if user is allowed to access the page
      *               0 : if user is NOT allowed to access the page
      */
-    public function page($p, $checkAction = false)
+    public function page(string $p, bool $checkAction = false)
```
Comment (abuathier) [inline centreon/www/class/centreonACL.class.php:1454]:
```suggestion
    public function page(string $p, bool $checkAction = false): int
```

File: centreon/www/class/centreonACL.class.php [hunk 6]
```diff
@@ -1445,25 +1445,28 @@ public function queryBuilder($condition, $field, $stringlist)
...
      * @return int | 1 : if user is allowed to access the page
      *               0 : if user is NOT allowed to access the page
      */
-    public function page($p, $checkAction = false)
+    public function page(mixed $p, bool $checkAction = false): int
     {
         $this->checkUpdateACL();
         if ($this->admin) {
-            return 1;
+            return self::ACL_ACCESS_READ_WRITE;
```
Comment (callapa) [inline centreon/www/class/centreonACL.class.php:1458]: Add constant

File: centreon/www/install/sql/centreon/Update-DB-23.04.0.sql [hunk 7]
```diff
@@ -0,0 +1 @@
+ALTER TABLE topology ADD topology_url_substitute VARCHAR(255) NULL AFTER topology_url_opt;
```
Comment (callapa) [inline centreon/www/install/sql/centreon/Update-DB-23.04.0.sql:1]: change the file version: 23.04.1

File: centreon/www/include/common/common-Func.php [hunk 8]
```diff
@@ -1552,21 +1552,31 @@ function get_child($id_page, $lcaTStr)
...
-                ORDER BY topology_order, topology_group ";
+        $rq = <<<SQL
+            SELECT topology_parent, topology_name, topology_id, topology_url, topology_url_substitute,
+                topology_page, topology_url_opt, is_react
+            FROM topology
+            WHERE topology_page IN ($lcaTStr)
+                AND topology_parent = :id_page
+                AND topology_page IS NOT NULL
+                AND topology_show = '1'
+            ORDER BY topology_order, topology_group
```
Comment (a-launois) [inline centreon/www/include/common/common-Func.php:1563]: Is this OK with mysql8 ?

Comment (callapa) [reply centreon/www/include/common/common-Func.php:1563]: Yes, otherwise the legacy pages would not be displayed at all

**Review by abuathier — APPROVED** [review]

**Review by dmyios — APPROVED** [review]

**Review by a-launois — APPROVED** [review]

**Pull request merged**